### PR TITLE
Koa Auto-Instrumentation Example

### PR DESCRIPTION
**Pull request by sm7x**
Auto-instrumentation example for Koa (to close the loop)

### CODE REVIEW

File: signalfx-tracing/signalfx-nodejs-tracing/koa/package.json [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  "dependencies": {
+    "dotenv": "^7.0.0",
+    "koa": "^2.7.0",
+    "koa-bodyparser": "^4.2.1",
+    "koa-logger": "^3.2.0",
+    "koa-router": "^7.4.0",
+    "koa-static": "^5.0.0",
+    "mongoose": "^5.4.22",
+    "node-wordnet": "^0.1.12",
+    "signalfx-tracing": "0.0.4",
```
Comment (rmfitzpatrick) [inline signalfx-tracing/signalfx-nodejs-tracing/koa/package.json]: `0.0.6` is the most recent, and this should work w/ any later maintenance release: https://docs.npmjs.com/about-semantic-versioning

File: signalfx-tracing/signalfx-nodejs-tracing/koa/server.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
+if (process.env.NODE_ENV !== 'production') {
+  require('dotenv').config();
```
Comment (rmfitzpatrick) [inline signalfx-tracing/signalfx-nodejs-tracing/koa/server.js]: Is this used?  I don't see a config file.

Comment (sm7x) [reply signalfx-tracing/signalfx-nodejs-tracing/koa/server.js]: Updated. Was for the previous API I used

File: signalfx-tracing/signalfx-nodejs-tracing/koa/package.json (hunk 1 above)
Comment (sm7x) [reply signalfx-tracing/signalfx-nodejs-tracing/koa/package.json]: Updated

File: signalfx-tracing/signalfx-nodejs-tracing/koa/README.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+## Using
+
+The WordExplorer allows you to add, retrieve, update and delete entries as part of a your vocabulary list. Each
+client request will automatically create an initiating parent span for distributed propagation
+to some basic REST api endpoints implemented via [Koa](https://koajs.com).
+
+
+
+TODO: Include example output here
```
Comment (achandras) [inline signalfx-tracing/signalfx-nodejs-tracing/koa/README.md]: Looks like you already have some sample output here. Is this TODO still needed?

Comment (sm7x) [reply signalfx-tracing/signalfx-nodejs-tracing/koa/README.md]: Thanks for catching that!

File: signalfx-tracing/signalfx-nodejs-tracing/koa/package.json [hunk 4]
```diff
@@ -0,0 +1,27 @@
+{
+  "name": "wordExplorer",
+  "version": "1.0.0",
+  "description": "A SignalFx Koa instrumentation example",
+  "main": "index.js",
+  "scripts": {
```
Comment (rmfitzpatrick) [inline signalfx-tracing/signalfx-nodejs-tracing/koa/package.json]: Can you add and document a mongo container creation command?

File: signalfx-tracing/signalfx-nodejs-tracing/koa/package.json [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+  },
+  "dependencies": {
+    "koa": "^2.7.0",
+    "koa-bodyparser": "^4.2.1",
+    "koa-logger": "^3.2.0",
+    "koa-router": "^7.4.0",
+    "mongoose": "^5.4.22",
+    "node-wordnet": "^0.1.12",
+    "signalfx-tracing": "0.0.6",
+    "wndb-with-exceptions": "^3.0.2"
```
Comment (rmfitzpatrick) [inline signalfx-tracing/signalfx-nodejs-tracing/koa/package.json]: not seeing yargs

File: signalfx-tracing/signalfx-nodejs-tracing/koa/README.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+            The server is listening on http://localhost:3000.
+
+  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+
+  Database connection established
+
+```
+From a different terminal, you may run the client commands.
+```bash
+  $ npm run client -- --help
```
Comment (rmfitzpatrick) [inline signalfx-tracing/signalfx-nodejs-tracing/koa/README.md]: `npm run client help` should be ok

File: signalfx-tracing/signalfx-nodejs-tracing/koa/package.json (hunk 5 above)
Comment (sm7x) [reply signalfx-tracing/signalfx-nodejs-tracing/koa/package.json]: Fixed

File: signalfx-tracing/signalfx-nodejs-tracing/koa/package.json [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    "start": "SIGNALFX_TRACING_DEBUG=true SIGNALFX_SERVICE_NAME=wordExplorerServer node ./server.js",
+    "client": "SIGNALFX_SERVICE_NAME=wordExplorerClient node ./client.js",
+    "mongo": "docker run -d -p 27017:27017 -v ~/data:/data/db mongo",
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "keywords": [],
+  "author": "SignalFx, Inc.",
+  "license": "ISC",
+  "devDependencies": {
+    "nodemon": "^1.18.10"
```
Comment (rmfitzpatrick) [inline signalfx-tracing/signalfx-nodejs-tracing/koa/package.json]: required?

File: signalfx-tracing/signalfx-nodejs-tracing/koa/wordExplorer/client.js [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+  let data = '';
+  res.on('data', (chunk) => {
+    data += chunk;
+  });
+  res.on('end', () => {
+    resolve(JSON.parse(data));
+  });
+}
+
+async function addWord(word, usage='') {
```
Comment (rmfitzpatrick) [inline signalfx-tracing/signalfx-nodejs-tracing/koa/wordExplorer/client.js]: does this need to be async?

Comment (sm7x) [reply signalfx-tracing/signalfx-nodejs-tracing/koa/wordExplorer/client.js]: Nope. Fixed. Thanks!

File: signalfx-tracing/signalfx-nodejs-tracing/koa/package.json (hunk 7 above)
Comment (sm7x) [reply signalfx-tracing/signalfx-nodejs-tracing/koa/package.json]: Fixed.

File: signalfx-tracing/signalfx-nodejs-tracing/koa/README.md [hunk 9]
```diff
@@ -0,0 +1,85 @@
+# Koa 2 Auto-Instrumentation Example
+
+This is an example of automatically producing distributed traces using the
+[SignalFx Tracing Library for JavaScript](https://github.com/signalfx/signalfx-nodejs-tracing).
+Please examine the instrumented [client](./client.js) and [server](./server.js) for
+some basic patterns in accessing the instrumentations of a [http](https://nodejs.org/api/http.html)
+client and a [Koa](https://www.koajs.com) application.
+
+In ths example, we have a simple vocabulary exploration application, named "WordExplorer",
```
Comment (rmfitzpatrick) [inline signalfx-tracing/signalfx-nodejs-tracing/koa/README.md]: `this example`

File: signalfx-tracing/signalfx-nodejs-tracing/koa/client.js [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+  console.log('\n++++++++++++++++++++++++++++++++++');
+  console.log();
+}
+
+yargs
+    .usage('Usage: $0 <command> [options]')
+    .scriptName('wordExplorer')
+    .command('add [word] [usage]', 'Add word to vocabulary list.',
+        {}, (argv) => {
+          client.addWord(argv.word, argv.usage).then((response) => {
```
Comment (rmfitzpatrick) [inline signalfx-tracing/signalfx-nodejs-tracing/koa/client.js]: nit: you can just use `then(printWord)`

Comment (sm7x) [reply signalfx-tracing/signalfx-nodejs-tracing/koa/client.js]: Updated

File: signalfx-tracing/signalfx-nodejs-tracing/koa/wordExplorer/mongoDriver.js [hunk 11]
```diff
@@ -0,0 +1,67 @@
+const MongoClient = require('mongoose');
+const MONGO_URL = `mongodb://localhost/words_db`;
+const mongoose = require('mongoose');
```
Comment (rmfitzpatrick) [inline signalfx-tracing/signalfx-nodejs-tracing/koa/wordExplorer/mongoDriver.js]: already imported as `MongoClient`

**Pull request merged**